agent_build_refactored/managed_packages/scalyr_agent_python3/agent-python3-config

```diff
@@ -0,0 +1,193 @@
+        preferred_openssl = PREFERRED_OPENSSL_FILE_PATH.read_text().strip()
+
+    if not preferred_openssl:
+        preferred_openssl = "auto"
+
+    if preferred_openssl == "auto":
+        try:
+            print("Looking for system OpenSSL >= 3", file=sys.stderr)
+            if look_for_system_openssl(OPENSSL_3_VERSION):
+                return
+
+            print("Looking for system OpenSSL >= 1.1.1", file=sys.stderr)
+            if look_for_system_openssl(OPENSSL_1_1_1_VERSION):
+                return
+        except Exception as e:
+            print(
+                f"Warning: Could not determine system OpenSSL version due to error: {str(e)}",
+                file=sys.stderr
+            )
+
+    print("Use embedded OpenSSL.", file=sys.stderr)
+    version_to_use = DEFAULT_FALLBACK_OPENSSL_VERSION
+    # Could not find any valid system OpenSSL, fallback to embedded one.
+    link_bindings(version_type=version_to_use, bindings_type="patched")
+
+    embedded_openssl_version = get_current_openssl_version(version_type=version_to_use)
+
+    print(f"Using embedded OpenSSL == {embedded_openssl_version}")
+
+
+def look_for_system_openssl(version_type: str) -> bool:
+    """
+    Check if OpenSSL library with given version is presented in system.
+    :param version_type: version type to use.
+    :return:
+    """
+    link_bindings(version_type=version_type, bindings_type="original")
+
+    system_openssl_version = get_current_openssl_version(version_type=version_type)
+
+    if not system_openssl_version:
+        return False
+
+    print(f"    Using system OpenSSL {system_openssl_version}", file=sys.stderr)
+    return True
+
+
+def link_bindings(version_type: str, bindings_type: str):
+    """
+    Configure Python's 'ssl' and 'hashlib' modules C bindings by linking to their
+        a particular shared objects.
+    :param version_type:  version of the OpenSSL to use, 1_1_1 or 3
+    :param bindings_type: Type of the C bindings. Can be:
+        'original': C bindings that look for system OpenSSL.
+        'patched': C bindings that are patched to use embedded OpenSSL library.
+    :return:
+    """
+
+    openssl_version_path = OPENSSL_LIBS_PATH / version_type / "bindings" / bindings_type
+
+    for binding_filename in [
+        SSL_C_BINDING_FILENAME,
+        HASHLIB_C_BINDING_FILENAME,
+    ]:
+        symlink_path = PYTHON_LIB_DYNLOAD_PATH / binding_filename
+
+        if symlink_path.exists():
+            symlink_path.unlink()
+
+        binding_path = openssl_version_path / binding_filename
+        symlink_path.symlink_to(binding_path)
+
+
+def get_current_openssl_version(version_type: str):
+    """
+    Determine current system OpenSSL version, if presented.
+    :param version_type: version of the OpenSSL to use.
+    :return:
+    """
+    result = subprocess.run(
+        [
+            sys.executable,
+            "-c",
+            "import ssl; print(ssl.OPENSSL_VERSION);"
```

Comment:
So we actually determine which OpenSSL is available on the system by checking against which OpenSSL version Python interpreter is compiled / linked against?

Does Python interpreter always dynamically links to OpenSSL library so this will always work as expected, or?

Reply:
Yes, by default, its C bindings for `_ssl` and `_hashlib` modules dynamically link to system's OpenSLL, but we also now. ship with the package another variants of those bindings that are parched to be "hardocded" to use OpenSSL
shared libraries that are also shipped with package in the `/opt/scalyr-agent-dependencies/lib/openssl` library.

I've also tried to add this to this docstring 
https://github.com/scalyr/scalyr-agent-2/blob/d48298a45a322219275048948038afdb0968664f/agent_build_refactored/managed_packages/managed_packages_builders.py#L50

Reply:
Thanks for adding this additional explanation.

I assume we have some post build / CI checks that ensure that the compiled Python binary was linked against the correct OpenSSL version?

Reply:
ah, thats a good idea.
I added that checks for some distributions. 
https://github.com/scalyr/scalyr-agent-2/blob/83b963f6d04596fbdac4c25344c7efc65d8c7bad/tests/end_to_end_tests/managed_packages_tests/test_packages.py#L484-L492

I think adding checks for all tested distributions is a bad idea since their versions will eventually change which will be causing additional flakiness of the tests.

full commit: https://github.com/scalyr/scalyr-agent-2/commit/e18d38e4af333d1abd3b1ec5c706e4bc607354cd